### ISSUE TITLE
Introduce "this" for default task results

### ISSUE THREAD

**Issue by dagwieers** [ansible/proposals#74]
# Proposal: Make task results available as "this" variable
*Author*: Dag Wieers <@dagwieers>

*Date*: 2017/09/20

- Status: New
- Proposal type: core design
- Targeted Release: 2.5
- Associated PR: https://github.com/ansible/ansible/pull/30957
- Estimated time to implement: 30 mins (already done)

## Motivation
We would like to make it easier for users to influence when the task failed or changed.

### Problems
- Registering return values that you don't need to expose to other tasks may overwrite an existing variable inadvertently
- It's an additional unnecessary statement
- Simplifies using **changed_when:**, **failed_when:** and **until:** for loop control

## Solution proposal
I would like to introduce a reserved variable name **this** that would consist of the return values of the specific task.

```yaml
  # Fail task when both files are identical
  - raw: diff foo/file1 bar/file2
    failed_when: this.rc == 0

  # Report a change when rc is not 2
  - shell: /usr/bin/billybass --mode="take me to the river"
    changed_when: this.rc != 2

  # Wait until file exists
  - stat:
      path: /tmp/some.file
    until: this.stat.exists

  # Wait until request works
  - uri:
      url: https://foo.bar/api/v1/cluster-manager/initial-install
      method: POST
      body:
        { "foo": "bar" }
      status_code: [ 202 ]
    until: this is successful
    retries: 10
    delay: 30
    delegate_to: localhost
```

This avoids having to needlessly register it to a specific (and unique) variable name, makes it very obvious where the values are coming from (without having to check what it was registered to).

It also makes the YAML file more readable with lines like: `until: this is successful`

**Note:** In case the user has registered a variable `this` already, the implementation will warn the user that it is overriding a reserved variable, but it will not touch it, so existing uses of `this` (if any) will keep on working. The warning allows the user to fix those cases.

## Testing (optional)
We could extend the existing tests to also test **this**.

## Documentation (optional)
A few examples could be added as part of the **failed_when**, **changed_when** and loop section.

**Comment by bcoca** [comment]
it is also interesting for controlling loop progression:

```
when: not (this.results[-1]|failed and this.results[-1]|skipped)
```

**Comment by dagwieers** [comment]
So first implementation done in https://github.com/ansible/ansible/pull/30957.
Please review.

**Comment by amenonsen** [comment]
I like both the proposal and the implementation (which I just applied locally and tried out for a while).

**Comment by amenonsen** [comment]
I wonder if 'result' wouldn't be a better name than 'this', but I don't feel strongly about it. I just think ``failed_when: result.rc != 42`` scans more easily.

**Comment by jhawkesworth** [comment]
I think 'this' is preferable to result as 'result' is sometimes used in examples for registered variables, and 'results' is used when registering loop results ( see http://docs.ansible.com/ansible/latest/playbooks_variables.html#registered-variables ) so it would be better to have a name this is distinct from 'result'.

I recall someone proposing something similar in the past and it was declined on the grounds of being too 'programmery'  i.e. that it would reduce playbook readability for non programmers.  If I recall though the suggestion then was to use something like '$_' which is definitely 'programmer magic' and bad for readability.

I would like a better word than 'this' that communicated the nicely-limited scope of such a variable, but cannot think of a better English word.

**Comment by rcarrillocruz** [comment]
+1 to this proposal, it striked when I started using Ansible why I had to manually register variables, it seems natural you want to do things per task outcome.

For reference, what @jhawkesworth refers about automatic 'results' on loops is described on http://docs.ansible.com/ansible/latest/playbooks_loops.html#using-register-with-a-loop .

I think precisely because 'result' is used in examples, it would be better to stick to that, for easier transition on already existing playbooks (I bet most variables used in the wild use 'result' with register) and it's more 'ansibly'. 'this' makes me remember Java :/ .

So, 'result' for a single task outcome as singular and 'results' as multiple tasks outcome as plural fits better IMHO.

**Comment by dagwieers** [comment]
I am not in favor of naming this **result**, for more than one reason:
- Chances of conflicts with existing playbooks is going to be very real, going to cause issues
- It does not clearly indicate the local scope
- Also **result** is often used for the *previous* task result
- We are not going to add a second **results** variable for loops since it is already available from the task result. So I prefer **this.results** over **result.results**, keep it simple

I don't mind a better name, but **result** is not the better name.

**Comment by bcoca** [comment]
I like the direction, just one caveat. We made a rule that any new 'ansible variables' need to be `ansible_` prefixed.

**Comment by amenonsen** [comment]
I think it would be better to not have the feature at all than to have it be named ``ansible_this``.

**Comment by kustodian** [comment]
I agree with @amenonsen. `this` sounds so natural and easy, while `ansible_this` just feels wrong.

**Comment by maxamillion** [comment]
+1 to the change.

I think `this` is a good choice, if someone wanted to make it more "specific" to ansible somehow, or even a playbook/taskset it could be something like `this_task`.

**Comment by caphrim007** [comment]
+1 this would be useful in test situations (in my use case writing integration tests) so that I could eliminate all the `register: result`'ing I do to assert changes.

**Comment by sivel** [comment]
Copying my opinion here.  I am -1 on the feature as a whole.  We try to do less magic, and be more explicit.  This is a magical feature in rough alignment with `_` perl, and I prefer being more explicit.

To re-iterate an argument from bcoca, this actually makes reading a playbook less obvious, and potentially makes the lives of playbook linters harder.

**Comment by nitzmahone** [comment]
^-- if we end up following that logic, this arguably should be considered in `loop` as well (eg, making the user explicitly name the loop var instead of `item`)... Though I land on the side of `this` is a good thing, we should probably be consistent in the design of a new language feature, whichever way we go.

**Comment by thaumos** [comment]
@sivel, just to expand on what you mean for those who don't understand.  You are suggesting that if a user wants to do anything with a returned result, they should register a variable and have a follow up task to use that.  Correct?

**Comment by abadger** [comment]
I'm a strong -1 to anything that's not namespaced.  it will break existing playbooks and be a trip hazard for new playbooks.  So "this" earns a strong -1 from me.

bcoca proposed in the meeting using "_" as a namespace and I'm okay with that.  So the variable could be "_this".  Dunders were also proposed.  I could also be okay with that. ("__this__").

thaumos brought up that being too "programmery" is a bad thing.  I somewhat agree with that but I think that if that's the case, then "this" is also programmery (and programmery with a programming language that isn't natural to ansible).... The only proposal that satisfies the namespace argument and the non-programmery aspect that I can think of is ansible_results.  (or perhaps better: ansible_task_results).  I suppose this could be a case where we ask whether it's more important to be readable ("ansible_task_results") or writable ("_this").  I think that the compromises in between these two extremes have all of the drawbacks of both sides without the benefits.

**Comment by tima** [comment]
+1 to `this' -- I think its a great idea. 

-1 to `_this` -- that is a programming idiom bleeding thru where it shouldn't be. Ansible positions its playbook as human readable/meaningful that don't require programming skills. Programming idioms like _ should not be bleeding into there. I used to be a perl programmer and hated that crap. I also hated `_meta` in the Ansible inventory. I'm hoping that dies as inventory plugins become the norm. 

We need to do more magic and do more convention over configuration. That is what customers want -- to make their lives easier getting stuff done and not prescribing to someone's sense of code purity.

**Comment by sivel** [comment]
> @sivel, just to expand on what you mean for those who don't understand. You are suggesting that if a user wants to do anything with a returned result, they should register a variable and have a follow up task to use that. Correct?

Yes, I am in favor of the current behavior.  If you want to do anything with the returned result, you should explicitly register a variable.

Do less magic, be more explicit.  I will always land on the side of being explicit.

**Comment by samdoran** [comment]
+1 to the feature and using `this` as the var name.

I understand the desire for less magic, however this feature change removes some tedium and friction for the playbook author. That's a good thing.

I also see that `_this` would be far less problematic a name, but that forces a programming convention on playbook authors, and no other "magic" or built-in variable we have has an underscore prefix. I think that would be taking the style of Ansible in the wrong direction.

**Comment by thaumos** [comment]
+1 to `this`.

I echo everything that @samdoran has just eloquently said.  Less code lines == less friction.

**Comment by abadger** [comment]
I'm strongly enough against "this" that I'll do what I can to veto it (do what I can because we don't actually have a veto power...)

**Comment by bcoca** [comment]
-1 to `this` as it breaks actual playbooks i know of.

If we implement this feature (i'm 1/2 of a mind with @sivel that explicit > magic for auditability) it has to be namespaced, i prposed `_this` as a compromise, but in the end the 'correct' one would be `ansible_task_result` as @abadger notes as this follows existing convention, not only in naming but as how we expose many 'magic vars' to the users.

**Comment by samdoran** [comment]
To do anything beyond the task, you have to register. This `this` is very short-lived — just a _tiny_ bit of magic.

But if `this` is problematic, then `ansible_task_result` is the most logical name. Less elegant, but also less problematic.

**Comment by bcoca** [comment]
it does not solve it, it annoys the user that is already using it and prevents them from using the new feature

**Comment by thaumos** [comment]
I am fine with `ansible_task_result` as well.

**Comment by dagwieers** [comment]
Well, if you don't want to annoy users, you shouldn't be rewriting loops either :-)

**Comment by bcoca** [comment]
@dagwieers that is actually trying to remove annoyance and confusion, if you think we accomplish the opposite, we can revisit it

**Comment by dagwieers** [comment]
@bcoca: Same for this proposal, for 99.99% of the users there's no impact.

**Comment by maxamillion** [comment]
+1 to @abadger in respect to namespaces. I think `ansible_task_results` is a good option, it's clear and concise. A Playbook reader will have a good hint at what that variable means by looking at it even if they have no prior knowledge of the feature. I do like `this` because it's short and a relatively well known idiom for programmers but I do agree with the arguments that it's venturing too much into the programmer world and could be off-putting or confusing to some.

**Comment by tima** [comment]
Right @dagwieers. I mean how far are we supposed to impeded progress in making automation with ansible easier and more frictionless for the overall user community because of what some seemingly small number of users may or may not have done in their playbooks? There are probably some users out there that used ansible_results and ansible_task_results out there already.

**Comment by bcoca** [comment]
@tima the `ansible_` namespace is reserved and we made the rule to make all new 'magic vars' use it, we are even moving from old vars `play_hosts` to `ansible_play_hosts` to keep this pattern, anyone using `ansible_` in their vars should already be aware of their special nature. `this` and other names are not in the same category.

**Comment by dagwieers** [comment]
@bcoca I think the `ansible_` namespace is an unfortunate decision, because it makes everything a lot more verbose for no good reason IMO. 'ansible_facts' is unfortunate, I think a reserved variable 'facts' (or another idiom/term) is preferable. Since that will become the namespace for all facts in Ansible there's no need to add ansible_ in front.

Granted it is unfortunate that we want to redesign things in Ansible this late in the game, but making things less convenient and intuitive because there may be a conflict with what people are using, is a problem in itself standing in the way of nice design. Harming all future users for a limited set of existing users.

If we were designing a new language nobody would be considering `ansible_task_results`, but we're blinded by concerns from the past to make a good judgment for the future IMO.

(Consider the python language that namespaces all reserved keywords with `python_` *shiver*)

**Comment by abadger** [comment]
While I mention "ansible_task_results" as being the logical conclusion of "non-programmery" it has one fatal flaw... it takes more typing to type that then "register: this".  So it's not going to be a usability improvement.  ("ansible_result" breaks even.  "_result" could also be a compromise between namespaced and less programmery than "this")

**Comment by maxamillion** [comment]
Ultimately I like the `ansible_` namespacing of things, I know it's more verbose and we all hate typing but it's removes the guess work of "where did that variable come from?" when reading a playbook you didn't write for at least the specific classification of variable that is "ansible built-in" but I'm open to other opinions on the topic.

**Comment by dagwieers** [comment]
@maxamillion I am not comparing with or without `ansible_`, but adding `ansible_` to everything compared to more sensible reserved dictionaries as namespaces. Names that don't need `ansible_`.

When we decided to namespace facts with `ansible_` it was for different reasons (and the convenience you're referring to). Today we have better ways to avoid conflicting names, there's no need anymore to prepend things with `ansible_` to make stuff 'unique".

**Comment by bcoca** [comment]
@dagwieers I disagree on the account we still get users confused and frustrated with the few 'non namespaced' variables we have out there (inventory_file, play_hosts, etc). the reason behind `ansible_` prefixing was to avoid user confusion and frustration, which it has done pretty good job with as the count of queries of that nature have almost disappeared.

**Comment by maxamillion** [comment]
@dagwieers In the current line of discussion I'm not concerned about uniqueness from the perspective of ansible, I'm concerned about it from the perspective of an user. In the case of the user, I like the explicit "things that start with `ansible_` come from ansible" which as an universal rule (in my opinion anyways) 
 keeps the learning curve low and maintains a level of consistency that aids in over all user experience.

**Comment by dagwieers** [comment]
@maxmillion Would be the same for a reserved `facts` dictionary, it comes from the system.

And what you say isn't really true, as some `ansible_` namespaced variables are actually set by users. And there's no guarantee at all. Which there would be for reserved keywords.

**Comment by maxamillion** [comment]
@dagwieers users deciding to set `ansible_` variables does provide confusion, but if we're going to change a bunch of stuff then why not make that truly reserved and not allow variables set by users to be in that namespace?

The point I was trying to make is that if there is a definition of "all variables that start with `ansible_` come from ansible itself" and the held true in practice, it would provide a pleasant consistency. Is it perfect? It is not, but I think it would be a net win.

An alternative, if it is the consensus that having a set of reserved words similar to how a programming language would, is to augment `ansible-doc` to provide that list upon request for quick reference as well as a listing in the documentation on the website.

I'm not strongly for or against either option, but I generally prefer namespacing and that holds true here as well.

**Comment by dagwieers** [comment]
Well, I don't like string-based namespaces. That's why I am pointing to using a dictionary instead.

**Comment by maxamillion** [comment]
@dagwieers that's fair.

**Comment by mgedmin** [comment]
Even if the accepted name is `ansible_task_result`, I could still define an alias `this: "{{ ansible_task_result }}"` in my `role/vars/main.yml` and use "`ailed_when: this.rc == 2` etc., because of Ansible's lazy evaluation, right?

**Comment by bcoca** [comment]
CLT disease, chronic lazy typer

**Comment by agaffney** [comment]
Good point. You can define the `this` variable for indirection once and use it all over the place.

**Comment by kustodian** [comment]
Or you can make this variable name (or its alias name) configurable in `ansible.cfg` to make more people happy.

**Comment by kustodian** [comment]
I'm all for this idea, any default variable will be better than nothing.

Regarding the name, if you are looking from the users' *understand-ability* perspective `ansible_task_result` is probably the best choice, but it's long and boring to write. On the other hand I'm not sure `this` is the best short choice, because if you're note a coder, you probably won't know about it. Maybe a better variable name would be just plain and simple `task`, e.g.:
```yml
  - raw: diff foo/file1 bar/file2
    failed_when: task.rc == 0

  - shell: cat /some/file
    changed_when: task.stdout != ""

  - stat:
      path: /tmp/some.file
    until: task.stat.exists
```

**Comment by agaffney** [comment]
Making the variable name configurable in `ansible.cfg` will probably lead to playbooks/roles that only work on a single machine. If someone wants to change the var, using another var for indirection is a better approach.

**Comment by amenonsen** [comment]
> Making the variable name configurable in `ansible.cfg` will probably lead to playbooks/roles that only work on a single machine. If someone wants to change the var, using another var for indirection is a better approach.

I agree, I see no need for configurability beyond @mgedmin's suggestion.

**Comment by nitzmahone** [comment]
This has been brewing for awhile... Given the extremely limited scope/lifetime of this var, I don't think its name needs to be configurable or namespaced, and we should be able to easily deal with conflicts.

If `this` is defined at the beginning of a task, issue a deprecation warning that the defined value will be masked by the implicit task result value in release X.Y.0 (and don't override it, providing maximum compatibility). In release X.Y.0, change the warning text to state that the var value has been masked by the implicit result value.

We could also scan the relevant post-exec expressions for usages of `this`, but that seems overkill IMO.

I also like @kustodian 's idea of `task` as the name better (less "programmery"?), but it also seems like it'd be more likely to conflict than `this`...

I'm still +1 for `this` (or `task`, or whatever), and it'd be nice to nail this down for 2.7 if we can reach some agreement.

**Comment by abadger** [comment]
I'm still -1 to non-namespaced variables.  We want to should breaking users
playbooks when any reasonable alternative exists.  Task and result are both
self documenting names but they are going to cause conflicts for people who
have written playbooks.

 _result is the variable I'd be for (with _task slightly behind. See my
critique of this later)  Easy to type and easy to understand that the
reason they need to be decorated is so they don't conflict with things the
user would write in a playbook normally.  It's programmer-y but it's the
level of programmer-y that a playbook author just already understand (ie:
that two variables named the same will conflict.  Therefore the names they
use need to be different than the names ansible allocates now and into the
future).

ansible_result, ansible_task, or ansible_task_result is in second place for
me as it is self-documenting but a pain to type.  Why can't we be usable
and readable out of the box by pretending with an underscore instead of
ansible_?

"this" is probably the most problematic name.  It is un-namespace so it
will conflict with some user's playbooks.  But it is not self explanatory.
To get an inkling of what it does you have to understand programming (and
only in certain languages... Neither python nor JavaScript programmers
would know what this refers to)  And even for programmers, it doesn't
reference an object from within the object's scope.  It references the
return value of the previous object.  If this were to map to programming
concepts, it would probably map to the task being executed, giving access
to interact and dynamically change the task as it's being executed.).

**Comment by webknjaz** [comment]
I'm with @abadger on this. `this` is ambiguous, makes programmers misunderstand its purpose, makes users wonder what `this` is. Also used in different contexts `failed_when`, somewhere in the next task, current task (`this`, right?) it doesn't make sense to me and just horrifies me if it's ever accepted like this. I mean `this`.

Totally agree with first and second points as well.

**Comment by webknjaz** [comment]
Oh, and talking about accessing such var from within current and previous task it seems like a good idea to have separate `ansible_current_task_result` and `ansible_previous_task_result`.

**Comment by dagwieers** [comment]
@webknjaz That is not the purpose of this proposal and overcomplicates matters. The whole point is that it *only* exists in the scope of the task. Please read the proposal before commenting.

**Comment by dagwieers** [comment]
Death by committee

**Comment by jamescassell** [comment]
tl;dr: @sivel is the only person entirely opposed to the feature, @abadger and @bcoca are -1 on `this` name, @maxamillion +0 on `this` name, everyone else is +1 to `this` name, though several say they'd be okay with `ansible_task_result` if it can't be `this`

Seems like there's demand for this feature and the biggest hurdle is the bikeshedding on the name.  The impetus behind the feature is convenience, hence the short `this` name.  The opposition to `this` is concerned with polluting the common namespace and breaking existing users (or preventing them from using the new feature.)  @mgedmin brings up that individuals could define `this: "{{ ansible_task_result }}"` in their role vars to use the short name but prevent ansible itself from polluting the namespace.

Here's my tally of the opinions above with respect to the `this` name:

| Person | Vote | Comments |
|--|--|--|
| sivel | -1 | "I am -1 on the feature as a whole." |
| abadger | -1 | on `this` name, wants `_result` or `_this`, but otherwise `ansible_result`, `ansible_task`, or `ansible_task_result` |
| bcoca | -1 | on `this` name |
| dagwieers | +1 | wrote proposal, strongly opposes `ansible_*` |
| maxamillion | +0 | sees both sides on `this`, okay with `ansible_task_result`, agnostic |
| samdoran | +1 | `ansible_task_result` is logical if it can't be `this` |
| thuamos | +1 | also okay with `ansible_task_result` |
| nitzmahone | +1 | |
| rcarrillocruz | +1 | |
| tima | +1 | thinks someone is probably already using `ansible_task_result` |
| caphrim007 | +1 | |
| mgedmin | | says he can just alias `this="{{ansible_task_result}}"`
| kustodian | | recommends `task` over `this`


FWIW, I'm strongly in favor of the `this` name (+1) as it is short and sweet.  Whatever the name, I think it's better to have this feature than not have it, and users can alias their own name if the chose name is too verbose.

**Comment by dagwieers** [comment]
Let me also remark that we have a protection that guarantees backward compatibility if `this` is already used. If that is the case we provide a warning and behave as before.

**Comment by samdoran** [comment]
Closing per https://github.com/ansible/ansible/pull/30957#issuecomment-702792803

**Comment by cognifloyd** [comment]
What follows might need to be a separate proposal. If that's the case I can file a proposal another day. In any case I need to jot my thoughts down so I don't forget.

I think the core team has essentially rejected that this bullet of the motivation / problem statement for this proposal:

> - It's an additional unnecessary statement

Without an explicit statement the feature is "too magical". Dropping the simplified `this` var also means that we can't "simplify" as described in this part of the problem statement:

> - Simplifies using changed_when:, failed_when: and until: for loop control

That leaves the kernel of the original problem statement:

> - Registering return values that you don't need to expose to other tasks may overwrite an existing variable inadvertently

Currently, re-registering with the same var name might result in warnings or force playbook or role authors to constantly come up with new names for the vars the registering just so they can use do/until or *_when conditions. This is still a significant UX issue that needs to be addressed.

So, given the requirement that addressing this problem requires an explicit opt-in from content authors, I think we can address this problem with something like the following:

### Alternative to magic `this` var

Add a new task key like `local_register` or `temp_register` that registers a var but only makes it available within the context of the current task. We need a good name the makes the scoped nature of the var clear.

Adjusting the original example to add `temp_register` we have:

```yaml
  # Fail task when both files are identical
  - raw: diff foo/file1 bar/file2
    temp_register: this
    failed_when: this.rc == 0

  # Report a change when rc is not 2
  - shell: /usr/bin/billybass --mode="take me to the river"
    temp_register: this
    changed_when: this.rc != 2

  # Wait until file exists
  - stat:
      path: /tmp/some.file
    temp_register: this
    until: this.stat.exists

  # Wait until request works
  - uri:
      url: https://foo.bar/api/v1/cluster-manager/initial-install
      method: POST
      body:
        { "foo": "bar" }
      status_code: [ 202 ]
    temp_register: this
    until: this is successful
    retries: 10
    delay: 30
    delegate_to: localhost
```

There's also a question on how this needs to handle loops.

```
  - command: "{{ cmd }}"
    loop:
      - /bin/true
      - /bin/false
      - /bin/true
    loop_control:
      loop_var: cmd
    temp_register: this
    when: this.results|last is successful
```

For a good experience with loops and loop control, we might need to do a register under loop_control

```
  - command: "{{ item }}"
    loop:
      - /bin/true
      - /bin/false
      - /bin/true
    loop_control:
      loop_var: cmd
      register_previous: last_cmd
    when: last_cmd is successful
```

**Comment by geerlingguy** [comment]
Just wanted to mention that this (heh, no pun intended) is the first time I noticed this proposal, but it does feel like the right kind of magic for an abstraction, and would save me a ton of keystrokes and having to register additional vars that persist in a playbook run.

I've always found it weird that a task can refer to it's own registered variable, as it seems like the `register` would happen at the final end of the task, after things like `changed_when` and `failed_when` had already been run (conceptually, that is), but I chalked it up as one of Ansible's odd behaviors and lived with it.

**Comment by bcoca** [comment]
@geerlingguy you can go even farther, in a loop, it references the previous iteration results (why you have to use |default for the first iteration).

I would want to revamp how register works and how a task can/should handle results, then adding this magic becomes simple (and even more desirable), but it is mostly a backwards incompatible change.

**Comment by cognifloyd** [comment]
> I would want to revamp how register works and how a task can/should handle results, then adding this magic becomes simple (and even more desirable), but it is mostly a backwards incompatible change.

A great candidate for ansible 3.0?

Though maybe there's a way to introduce the revamped results registration similar to how `loop` was introduced next to `with_*`. Any new keywords could have additional restrictions so they only add vars at carefully defined points in time.
```yaml
- command: /bin/true
  results: name_of_var_for_following_tasks  # not available in changed_when, failed_when, loop etc
```

And then in a follow up, introduce an opt in flag for `this`
```yaml
- command: /bin/false
  use_this: yes  # this is only available in changed_when, failed_when - if task is looped, only populated after all loop iterations complete
  failed_when: this_task.rc == 0
```

And an opt in option for controlling a loop
```yaml
- command: "{{ item }}"
  loop:
    - /bin/true
    - /bin/false
    - /bin/true
  loop_control:
    loop_when: last_iteration is succeeded  # only applied after the first loop iteration
```

**Comment by cognifloyd** [comment]
Next idea:
Let's expose the results of the current task through a `result()` jinja function instead of through a variable.
(`range()` and `dict()` are builtin jinja functions; `lookup()`, `q()`, and `now()` are ansible-defined jinja functions).

This is just like the `result()` function in [StackStorm orquesta workflows](https://docs.stackstorm.com/orquesta/start.html#authoring) (example docs show YAQL, but its Jinja support is nearly equivalent - just replace `<% %>` with `{{ }}`).

Modifying the original example to use a `result()` function instead of `this` magic var:
```yaml
  # Fail task when both files are identical
  - raw: diff foo/file1 bar/file2
    failed_when: result().rc == 0

  # Report a change when rc is not 2
  - shell: /usr/bin/billybass --mode="take me to the river"
    changed_when: result().rc != 2

  # Wait until file exists
  - stat:
      path: /tmp/some.file
    until: result().stat.exists

  # Wait until request works
  - uri:
      url: https://foo.bar/api/v1/cluster-manager/initial-install
      method: POST
      body:
        { "foo": "bar" }
      status_code: [ 202 ]
    until: result() is successful
    retries: 10
    delay: 30
    delegate_to: localhost
```

`result()` has the benefits of:

1. it is explicit/opt-in,
2. it is NOT a magic var, so the `ansible_` prefix is not required,
3. it is almost as short as typing `this` (8 chars vs 4 chars),
3. it dodges the issue of variable precedence since it is a function, and
4. it is backwards compatible even if someone has a var named `result` because `result()` is a function call, not a variable access.

It might, however, conflict with any custom filters named `result`. Is that actually a concern?

**Comment by bcoca** [comment]
so, i created this due to some questions about the internal aspects on how ansible templates (or should template) things internally, only the bottom 2 sections are relevant for this discussion, but might clarify where such a keyword would be needed and how it needs to work:

```
# format:
# keyword: *dependency <templating priority>
# _ indicates alias

#####################
##### pre execution
#####################

collections: 100
vars: 100

_templ: *vars, *collections

name: *_templ 90
name: *_tags 90

async: *_templ 80
poll: *_templ 80
run_once: *_templ 80
throttle: *_templ 80

loop: *_templ 70
loop_control: *_templ 70
with_: *_templ 70

_l: *loop, *loop_control

when: *_l 60

no_log: *_l 50

retry: *_l 40
delay: *_l 40
timeout: *_l 40

check_mode: *_l 35
delegate_to: *_l 35
debugger: *_l 35
diff: *_l 35
environment: *_l 35
module_defaults: *_l 35

#######################
#######  task executes
#######################

action: *_l 30
local_action: *_l 30

become: *_l 25
become_exe: *_l 25
become_flags: *_l 25
become_user: *_l 25
become_method: *_l 25
connection: *_l 25
remote_user: *_l 25

_conninfo: become, become_exe, become_flags, become_user, become_method, connection, remote_user

args: *action, *conninfo 20

########################
####### task item result
#######################

register: !static *_item_result  #shows previus item in loop, requires |default for first item

delegate_facts: *_l, *register 5
failed_when: *_l, *register 5
changed_when: *_l, *register 5

notify: *register, *changed_when 0

########################
####### task final
#######################

register: !static *_results # if loop, has 'results' list with each item return in order, sumarized states accessible as normal

failed_when: *register 5
changed_when: *register 5

any_errors_fatal: *register, *failed_when 0
ignore_errors: *register, *failed_when 0
notify: *register, *changed_when 0
until: *register, *failed_when, *changed_when 0
```
